_posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md

```diff
@@ -0,0 +1,904 @@
+
+- **OpenTelemetry** for creation, propagation, collection, processing, and exporting of trace data.
+
+- **OpenSearch** as the sink for the traces
+
+OpenSearch is a community-driven, open source search and analytics suite derived from Apache 2.0 licensed Elasticsearch 7.10.2 & Kibana 7.10.2. Additionally, it comes with managed Trace Analytics plugin, a visualization and user interface, and OpenSearch Dashboards.
+
+- **Data Prepper** and **Jaeger** to re-format OpenTelemetry trace data and export it to OpenSearch in formats that both the OpenSearch dashboard and Jaeger understand. (This should go away soon as both Jaeger and OpenSearch Dashboards will natively be able to use OTLP traces)
+
+[Jaeger](https://www.jaegertracing.io/) tracks and measures requests and transactions by analyzing end-to-end data from service call chains, making it easier to understand latency issues in microservice architectures. It's exclusively meant for distributed tracing with a simple web UI to query traces and spans across services. Jaeger enables views of individual traces in a waterfall-style graph of all the related trace and span executions, service invocations for each trace, time spent in each service and each span, and the payload content of each span, including errors.
+
+Data Prepper is a new component of OpenSearch that receives trace data from the OpenTelemetry collector, aggregates, transforms, and normalizes it for analysis and visualization in OpenSearch Dashboards.
+
+- **Trace Analytics Plugin** for OpenSearch to visualize and query traces
+
+The Trace Analytics plugin also aggregates trace data into Trace Group views and Service Map views, to enable monitoring insights of application performance on trace data. By going beyond the ability to search and analyze individual traces, these features enable developers to proactively identify application performance issues, not just react to problems when they occur.
+
+## Architecture
+
+Once we got the components for the pipeline penciled, the remaining work to get them to play nice with each other and scale to our requirement. Since then, the tracing pipeline has been launched in production and processes 4-6 TB of traces daily. In a nutshell, the flow is as follows:
+
+**&#8594;** Applications use OpenTelemetry libraries/API to instrument traces and send it to open telemetry agents.
+
+**&#8594;** OpenTelemetry agents process/batches and send traces from microservices to the openTelemetry gateway collector.
+
+**&#8594;** The collector processes/samples the traces and export them to backends which in our case are data prepper and jaeger collector. These backends transform and export the data to OpenSearch.
+
+**&#8594;** Trace Analytics plugin for OpenSearch Dashboards and Jaeger visualizes trace data from OpenSearch and allows querying on it.
+
+![tracing_step4](/assets/media/blog-images/2021-12-02-distributed-tracing-pipeline-with-opentelemetry/tracing_step4.png){: .img-fluid}
```

Comment:
I see what you are saying about the image. I do think the last image with all the components has added verbosity which makes its slightly hard to read. Since I do mention it in the text, I think I can reduce the verbosity in the image to make it more clear/legible. I hope that works as I do like the images in 3D, Ill try to edit the last image and make it less verbose.

Reply:
I do think that removing some of the text on the labels would help. Some have text in parenthesis which is probably not necessary.

Also, can you change the perspective so that it is not as turned? It looks like it is almost a 45 degree angle. Maybe a 20-25 degree angle would help as well? I'm unsure of how your diagram tool works in this regard.

Reply:
So I removed some of the text and resized the image on step 4, it looks much easier to read. Also changed OpenTelemetry Collector to OTEL Collector. Unfortunately with cloudcraft.io, it doesn't allow me to set an angle for the 3D images so this is just what comes by default. 

If this looks good, then I can modify other images as well, to go from OpenTelemetry Collector to OTEL Collector

Reply:
> This {: .img-fluid} is rendering in the final HTML as displayed in GitHub. I believe this happens with all the images. I'm unsure how this renders in the final tooling though.

I'm unsure of this as well, @stockholmux might have a better idea since it was his suggestion.

Reply:
I've only looked in GitHub's rendered markdown using your repo:
https://github.com/kuberkaul/project-website/blob/main/_posts/2021-12-02-distributed-tracing-pipeline-with-opentelemetry.md

I didn't run Jekyll. It might not show up in the HTML which Jekyll renders.

Reply:
I took a look. I think it does look a lot better. If you change the other images, then they should all be more legible. Thanks!

Reply:
awesome and done ✔️

Reply:
The `{:img-fluid}` makes the images resize as the page scales. It shows up in github because of a different flavour of markdown. 

FYI - Spinning up an instance of the website is pretty quick it might help root out these little issues - that's how I review it.